app: Use relayFee setting

JoeGruffins opened the pull request:
app: Use relayFee setting.

~Make relayFee a field of the blockchain and set that field on initiation
and changing.~

Remove the realyFee method from dcrdata and have all relay fees passed
in order to separate functions cleanly. Relay fees now depend upon the
account calling the dcrdata method.

Follow up on #152

JoeGruffins committed:
account: Use relayFee setting.

Remove the realyFee method from dcrdata and have all relay fees passed
in order to separate functions cleanly. Relay fees now depend upon the
account calling the dcrdata method.

buck54321 commented on decred/decred/dcr/account.py:
```diff
@@ -1907,13 +1907,13 @@ def purchaseTickets(self, qty, price):
             poolAddress=pi.poolAddress,
             votingAddress=pi.ticketAddress,
-            ticketFee=0,  # use network default
+            ticketFee=DefaultRelayFeePerKb,
```
I noticed this is set separately in dcrwallet with `setticketfee`. We could 1) Leave it like it is. All tickets at the network default 10 atoms/byte rate, or 2) Use the user's `relayFee` setting, or 3) implement our own `setTicketFee`.

JoeGruffins replied:
When trying to use our relay fee, default and high were fine, but if I tried to use the lower 4 atoms/b setting, dcrdata threw an error back at me. So, either default or the `setticketfee` you mention.  Is default ok for this pr, since that wouldn't change anything, and we can investigate `setticketfee` in another issue/pr?

buck54321 commented on decred/decred/dcr/dcrdata.py:
```diff
@@ -861,7 +852,7 @@ def saveBlockHeader(self, header):
         self.headerDB[bHash] = header
 
-    def sendToAddress(self, value, address, keysource, utxosource, feeRate=None):
+    def sendToAddress(self, value, address, keysource, utxosource, relayFee):
```
If we're not going to have a default, let's add an optional argument with default value `allowHighFees=False` and then check that the fees aren't outrageous.

buck54321 replied:
Could you do a little work on the docstring here too?

buck54321 commented on decred/decred/dcr/dcrdata.py:
```diff
@@ -1373,7 +1363,7 @@ def purchaseTickets(self, keysource, utxosource, req):
         return (splitTx, tickets), splitSpent, internalOutputs
 
-    def revokeTicket(self, tx, keysource, redeemScript):
+    def revokeTicket(self, tx, keysource, redeemScript, relayFee):
```
Document the new arg.

buck54321 commented on decred/decred/dcr/dcrdata.py:
```diff
@@ -962,7 +953,7 @@ def confirmUTXO(self, utxo, block=None, tx=None):
         return False
 
-    def sendOutputs(self, outputs, keysource, utxosource, feeRate=None):
+    def sendOutputs(self, outputs, keysource, utxosource, relayFee):
```
Looks like the argument documentation was already missing, but please add it while you're here.

JoeGruffins committed:
Make relayFee an integer everywhere.

Because an atom cannot be divided, and we convert it to an integer
already in a few places, it should be an integer and not a float from
the beginning.

Also fix docstrings.

JoeGruffins commented:
I can't imagine why we would need the rate as a float. It is unneeded precision. We explicitly convert it int a couple places already. Can we just make it an int from the beginning?

buck54321 commented on decred/decred/dcr/account.py:
```diff
@@ -1865,7 +1865,7 @@ def addressSignal(self, addr, txid):
             self.signals.balance(self.calcBalance())
 
-    def sendToAddress(self, value, address, feeRate=None):
+    def sendToAddress(self, value, address):
```
This is fine for now. We may want to add this back as a way to override the default fee rate at some point.

buck54321 commented on decred/decred/dcr/dcrdata.py:
```diff
@@ -1125,6 +1143,9 @@ def purchaseTickets(self, keysource, utxosource, req):
                 UTXOs. The filterFunc is an optional function to filter UTXOs,
                 and is of the form func(UTXO) -> bool.
             req account.TicketRequest: the ticket data.
+            relayFee (int): Transaction fees in atoms per kb.
```
It's worth mentioning here that for the ticket (but not the split tx), the `relayFee` is overridden by the `req.txFee`.

JoeGruffins replied:
hmm, why have a relayFee parameter at all? Can we remove one of them?

buck54321 replied:
Like you discovered, the ticket fee and split tx fee may need different fees. If the user wants to pay the low fee on the split tx, you apparently need to set the ticket separately.

JoeGruffins replied:
I see. But what I mean is that we have the split tx fee defined in the req already, so we don't need it again in the constructor, which is overridden by the req.txFee. Maybe I'm not seeing something obvious here? We just need to two fee rates, which are in the req already.

buck54321 commented on decred/decred/dcr/dcrdata.py:
```diff
@@ -1272,12 +1295,12 @@ def purchaseTickets(self, keysource, utxosource, req):
 
         # Send the split transaction.
         # sendOutputs takes the fee rate in atoms/byte
         splitTx, splitSpent, internalOutputs = self.sendOutputs(
-            splitOuts, keysource, utxosource, int(txFeeIncrement / 1000)
+            splitOuts, keysource, utxosource, txFeeIncrement // 1000, allowHighFees
```
This seems like it's a bug. `sendOutputs` doesn't actually take fees in `atoms / kB`. Shouldn't be coverting here. We've been sending split transactions with low fees, I think.

JoeGruffins replied:
Looks that way.

JoeGruffins replied:
Here,s a ticket purchase from a month ago done with tinydecred, but fee looks ok https://testnet.dcrdata.org/tx/0db01ee2735021b979f0a19ca5deca62d1ef615daf63dca0b32be33c62cdabc2/out/32

JoeGruffins replied:
Ah, I see. I removed it but, this was being converted back to coins inside of sendOutputs.

buck54321 replied:
You can go to atoms/kB everywhere below the `AccountSettingsScreen` if you'd like.

JoeGruffins replied:
I guess we could go all atoms/byte. I'll make an issue about it later. I think they are all in atoms/kB below the AccountSettingsScreen currently.

buck54321 reviewed (changes requested):
I think found a pre-existing bug.

JoeGruffins commented on decred/decred/dcr/dcrdata.py:
```diff
@@ -998,14 +989,13 @@ def sendOutputs(self, outputs, keysource, utxosource, feeRate=None):
         changeScriptSize = txscript.P2PKHPkScriptSize
 
-        relayFeePerKb = feeRate * 1e3 if feeRate else self.relayFee()
         for (i, txout) in enumerate(outputs):
-            checkOutput(txout, relayFeePerKb)
+            checkOutput(txout, relayFee)
 
```
It was converted here. So I guess it did take atoms/b. Should I revert? Or try to make everything atoms/b? Either is fine with me, but some accepting atoms/byte and other atoms/kb is confusing.

JoeGruffins committed:
Remove relayFee parameter from purchaseTickets.

The relay fee is already included in the req, which the caller has
defined. They can change the relay fee there.

buck54321 reviewed (changes requested):
This is working well, except now I'm getting rejected when broadcasting a regular transaction at the low fee. Looking at dcrdata output, I see 

```
2020-06-06 18:05:41.663 [ERR] PSQL: SendRawTransaction failed: -1: rejected transaction 59b6d9693deba6fdb84173fd97308fded12b0bc299986641eaf4aadb0c5cd710: transaction 59b6d9693deba6fdb84173fd97308fded12b0bc299986641eaf4aadb0c5cd710 has insufficient priority (2.8568972827586208e+07 <= 5.76e+07)
```

So the node won't even accept regular transactions at that rate. Could you check if you are seeing the same error, and if so we can readjust.
